plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/util/overlaps/RectangleStripOverlapRemover.java

```diff
@@ -0,0 +1,368 @@
+    }
+    
+    /**
+     * Applies the movement computed for the given rectangle node's rectangle to its original rectangle, paying
+     * attention to the overlap removal direction.
+     * 
+     * @param rectangleNode
+     *            the rectangle node whose results to apply to the original rectangle.
+     * @param stripSize
+     *            strip size computed by the overlap removal algorithm.
+     */
+    private void exportRectangle(final RectangleNode rectangleNode, final double stripSize) {
+        ElkRectangle rectangle = rectangleNode.rectangle;
+        ElkRectangle originalRectangle = rectangleNode.originalRectangle;
+        
+        switch (overlapRemovalDirection) {
+        case UP:
+            originalRectangle.y = startCoordinate - rectangle.height - rectangle.y;
+            break;
+            
+        case DOWN:
+            originalRectangle.y += startCoordinate;
+            break;
+            
+        case LEFT:
+            originalRectangle.x = startCoordinate - rectangle.height - rectangle.y;
+            break;
+            
+        case RIGHT:
+            originalRectangle.x = startCoordinate + rectangle.y;
+            break;
+        }
+    }
+
+
+    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+    // Actual Algorithm
+    
+    /**
+     * Triggers the overlap removal algorithm.
+     * 
+     * @return the height of the resulting label strip when removing horizontal overlaps, or the width of the resulting
+     *         label strip when removing vertical overlaps.
+     */
+    public double removeOverlaps() {
+        // Check if an overlap removal strategy is used; if not, fall back on the greedy strategy
+        if (overlapRemovalStrategy == null) {
+            overlapRemovalStrategy = new GreedyRectangleStripOverlapRemover();
+        }
+        
+        // Sort the list of rectangles
+        rectangleNodes.sort(RectangleStripOverlapRemover::compareLeftRectangleBorders);
+        
+        // Compute and remove overlaps
+        computeOverlaps();
+        double stripSize = overlapRemovalStrategy.removeOverlaps(this);
+        
+        // Apply the results
+        rectangleNodes.stream()
+                .forEach(node -> exportRectangle(node, stripSize));
+        
+        return stripSize;
+    }
+    
+    /**
+     * Populates the {@link RectangleNode#overlappingNodes} lists. This method uses a scanline algorithm that remembers
+     * which rectangles currently intersect the scanline. As it encounters a new rectangle, it first removes rectangles
+     * that do not intersect the scanline anymore and then adds overlaps between all currently intersecting rectangles.
+     */
+    private void computeOverlaps() {
+        SortedSet<RectangleNode> intersectingNodes = Sets.newTreeSet(
+                RectangleStripOverlapRemover::compareRightRectangleBorders);
+        double scanlinePos = Double.MIN_VALUE;
```

Comment:
Would if your positions were negative one day ... you never know. Otherwise `0` would make more sense instead of `Double.MIN_VALUE`.